### ISSUE TITLE
[REVIEW]: GBOML: Graph-Based Optimization Modeling Language

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4158]
**Submitting author:** @MiftariB (<a href="http://orcid.org/0000-0001-5334-0234">Bardhyl Miftari</a>)
**Repository:** https://gitlab.uliege.be/smart_grids/public/gboml
**Branch with paper.md** (empty if default branch): 
**Version:** v0.1.1
**Editor:** @fraukewiese
**Reviewers:** @odow, @leonardgoeke
**Archive:** 10.5281/zenodo.6470268

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/29a98e5ccf2778eb41c8b725a809e51b/status.svg)](https://joss.theoj.org/papers/29a98e5ccf2778eb41c8b725a809e51b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/29a98e5ccf2778eb41c8b725a809e51b"><img src="https://joss.theoj.org/papers/29a98e5ccf2778eb41c8b725a809e51b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/29a98e5ccf2778eb41c8b725a809e51b/status.svg)](https://joss.theoj.org/papers/29a98e5ccf2778eb41c8b725a809e51b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://gitlab.uliege.be/smart_grids/public/gboml" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@odow & @leonardgoeke, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @fraukewiese know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @odow
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://gitlab.uliege.be/smart_grids/public/gboml">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.uliege.be/smart_grids/public/gboml">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@MiftariB) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @leonardgoeke
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://gitlab.uliege.be/smart_grids/public/gboml">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.uliege.be/smart_grids/public/gboml">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@MiftariB) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @odow, @leonardgoeke it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1140

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.24 s (376.7 files/s, 63682.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          42           2574           1795           6915
JSON                             8              0              0           1735
reStructuredText                29            537            640            424
TeX                              1             18              0            176
Markdown                         4             40              0             87
YAML                             2              8              2             36
DOS Batch                        1              8              1             26
make                             1              4              7              9
TOML                             1              0              0              3
-------------------------------------------------------------------------------
SUM:                            89           3189           2445           9411
-------------------------------------------------------------------------------


Statistical information for the repository '0293c427244edeb7ca01327c' was
gathered on 2022/02/14.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
AdrienBolland                    1            21             22            0.12
Bardh                          140         17830          10443           80.85
Christophe Druet                 1             1              0            0.00
Hatim Djelassi                  20           690            471            3.32
Mathias Berger                   5          1449             23            4.21
MiftariB                        18          3163            855           11.49

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Bardh                     10853           60.9          4.8                6.62
Hatim Djelassi              103           14.9         11.1                0.00
Mathias Berger               82            5.7          0.5               46.34
MiftariB                    246            7.8          4.7                5.28
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04158/joss.04158/10.21105.joss.04158.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04158/joss.04158/10.21105.joss.04158.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1137/15M1020575 is OK
- 10.1007/s12532-011-0026-8 is OK
- 10.1007/s12532-012-0036-1 is OK
- 10.1287/mnsc.36.5.519 is OK
- 10.1007/s12532-009-0008-2 is OK
- 10.1007/978-1-4613-0215-5_8 is OK
- 10.5334/jors.188 is OK
- 10.23919/PSCC.2018.8442948 is OK
- 10.21105/joss.00825 is OK
- 10.1016/j.esr.2018.01.003 is OK
- 10.1016/j.enpol.2011.06.033 is OK
- 10.3389/fenrg.2021.671279 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by fraukewiese** [comment]
@odow @leonardgoeke  – This is the review thread for the paper. All of our communications will happen here from now on.

Please read the "Reviewer instructions & questions" in the first comment above. Both reviewers have checklists at the top of this thread (in that first comment) with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention https://github.com/openjournals/joss-reviews/issues/4158 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for the review process to be completed within about 4-6 weeks but please make a start well ahead of this as JOSS reviews are by their nature iterative and any early feedback you may be able to provide to the author will be very helpful in meeting this schedule.

**Comment by odow** [comment]
The repo is on a self-hosted GitLab instance, so linking won't work. Here's the start of a review issue: https://gitlab.uliege.be/smart_grids/public/gboml/-/issues/1

**Comment by fraukewiese** [comment]
Thanks @odow for opening the gitlab issue.

**Comment by MiftariB** [comment]
Dear Oscar and Frauke,

@odow : First and foremost, thank you for accepting to review this submission and providing valuable comments so quickly. We really appreciate the feedback and we are currently working to address it. We will provide a written answer shortly and make some adjustments in the paper as well.

@fraukewiese  : @odow highlighted several points that should be updated in our paper, such as a more in-depth comparison with other tools and providing a short example of GBOML code. We would be happy to do so and wanted to check with you whether the original 1000 word limit could be exceeded?

Best Regards, 

MIFTARI B

**Comment by fraukewiese** [comment]
@MiftariB : 1000 words is not a very strict limit. However please still be concise when doing the additions in the paper and maybe check if you can shorten at other parts.

**Comment by leonardgoeke** [comment]
I've added my comments to the review thread, but I don't manage to edit the checklist. My best guess is I'm missing the permission.

**Comment by fraukewiese** [comment]
@editorialbot add @leonardgoeke as reviewer

**Comment by fraukewiese** [comment]
@leonardgoeke : Sorry for the late reply and inconvenience. I have tried to assign you again. Could you try if you can edit the checklist now?

**Comment by kthyng** [comment]
Hi @leonardgoeke! We had a tech change for JOSS and now you can make your own checklist with: `@editorialbot generate my checklist`

@fraukewiese this is the way to do this going forward.

**Comment by leonardgoeke** [comment]
@editorialbot generate my checklist

**Comment by MiftariB** [comment]
Dear Oscar and Leonard,

We are pleased to let you know that the paper has now been updated. We tried to address your comments and implement the action items that we listed in the Gitlab issue as best we could. The only deviation from them was probably in the description of the SMS++ and Plasmo.jl frameworks. After putting some more thought into it, it seemed more appropriate and fair to us to focus on and concisely state the vision that underpins each tool, rather than getting into a detailed discussion of similarities, differences and limitations that may stem from their current development status.We would be happy to have your thoughts on the updated version of the paper.

Best,

GBOML Team

**Comment by MiftariB** [comment]
@editorialbot generate pdf

**Comment by MiftariB** [comment]
Dear @fraukewiese, 

Some elements are missing in the previewed pdf : 
- Co-authorship 
- Orcid of authors 

Is this normal for a preview version ? Is it possible to add an email address ? 

Best Regards, 

Miftari Bardhyl

**Comment by leonardgoeke** [comment]
My objections are sufficiently addressed in the revised version and I’ve updated my checklist above accordingly.

**Comment by fraukewiese** [comment]
@MiftariB : Yes, that is normal for a preview version. You can have a look at already published papers, there you can see that the Orcid are not directly included in the paper pdf but still available. About the email address: They are usually not added in the paper.

**Comment by fraukewiese** [comment]
@leonardgoeke Thank you very much for your thorough review!

**Comment by fraukewiese** [comment]
@odow : Thank you very much for your fast and thorough review. Are you satisfied with the changes made by the authors?

**Comment by odow** [comment]
Yes, I'm happy with the latest revision.

**Comment by MiftariB** [comment]
Dear @fraukewiese, 

Thanks for the precision concerning the paper. We are unsure of what is expected from us at this point. Could you clarify the next steps? 

Best Regards,

Miftari B

**Comment by fraukewiese** [comment]
@odow Thank you very much for your thorough review!

**Comment by fraukewiese** [comment]
@MiftariB 
At this point could you:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by MiftariB** [comment]
Dear @fraukewiese, 

Thanks for the instructions.
 - The version tag is "JOSS" :  https://gitlab.uliege.be/smart_grids/public/gboml/-/tags/JOSS
 - The archival deposit is : https://zenodo.org/record/6470268#.Yl6-k_NByeA
 - The DOI is : 10.5281/zenodo.6470268

Regards, 

MIFTARI B

**Comment by fraukewiese** [comment]
@editorialbot set 10.5281/zenodo.6470268 as archive

**Comment by fraukewiese** [comment]
@editorialbot set JOSS as version

**Comment by fraukewiese** [comment]
@MiftariB : Could you make a numbered version tag of the style vx.x.x ? Thank you.

**Comment by MiftariB** [comment]
@fraukewiese, the new tag is [v0.1.1](https://gitlab.uliege.be/smart_grids/public/gboml/-/tags/v0.1.1)

**Comment by MiftariB** [comment]
No particular reason, it was just to make it clearer which version was related to the JOSS publication. [v0.1.1](https://gitlab.uliege.be/smart_grids/public/gboml/-/tags/v0.1.1) is just fine.

**Comment by fraukewiese** [comment]
@editorialbot recommend-accept

**Comment by MiftariB** [comment]
@fraukewiese there is still the co-authorship missing in the [PDF](https://github.com/openjournals/joss-papers/blob/f8e18c51ea8c184deb4956ee59df17b288ef5a0f/joss.04158/10.21105.joss.04158.pdf). What could we do to fix that ?

**Comment by arfon** [comment]
> @fraukewiese there is still the co-first authorship missing in the [PDF](https://github.com/openjournals/joss-papers/blob/f8e18c51ea8c184deb4956ee59df17b288ef5a0f/joss.04158/10.21105.joss.04158.pdf). What could we do to fix that ?

Please see this recent example on how to denote co-authorship: https://gitlab.com/culturalcartography/text2map/-/blob/master/paper/paper.md

**Comment by MiftariB** [comment]
@arfon Thanks for the reply. However, I am unsure of what needs to be done. In the above mentionned example, they wrote : 
```
authors:
- name: Dustin S. Stoltz^[co-first author]
  orcid: 0000-0002-4774-0765
  affiliation: 1
- name: Marshall A. Taylor^[co-first author]
  orcid: 0000-0002-7440-0723
  affiliation: 2
```
We wrote : 
```
- name: Bardhyl Miftari^[co-first author, corresponding author]
    orcid: 0000-0001-5334-0234
    affiliation: 1
- name: Mathias Berger^[co-first author, corresponding author]
    orcid: 0000-0003-3081-4833
    affiliation: 1
```
And when compiling our paper on https://whedon.theoj.org/ everything seems fine. I am no expert but It seems like the new editorialbot does not accept having both **co-first author** and **corresponding author** as footnote at once.

**Comment by arfon** [comment]
Ah OK, I see what's going on here. @tarleb – any chance you could take a look here? It looks like the footnotes here are breaking with both 'co-first author' and 'corresponding author'.

**Comment by MiftariB** [comment]
@editorialbot generate pdf

**Comment by MiftariB** [comment]
@editorialbot generate pdf

**Comment by MiftariB** [comment]
We removed the "corresponding author" tag and everything works now on the preview. Can we proceed with this version of the paper ?

**Comment by tarleb** [comment]
I think the following should work:

``` yaml
authors:
  - name: Bardhyl Miftari
    orcid: 0000-0001-5334-0234
    equal-contrib: true
    corresponding: true
    affiliation: 1
  - name: Mathias Berger
    orcid: 0000-0003-3081-4833
    equal-contrib: true
    corresponding: true
    affiliation: 1
```

**Comment by MiftariB** [comment]
@editorialbot generate pdf

**Comment by MiftariB** [comment]
Everything seems good now. Thanks.

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@odow, @leonardgoeke – many thanks for your reviews here and to @fraukewiese for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@MiftariB – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by MiftariB** [comment]
Thank you @odow and @leonardgoeke for your reviews and your time ! Your feedback was very valuable to us. As the paper has been published, we've closed the issue on our gitlab but feel free to write to us for any question or remark. Thanks to @fraukewiese for editing this paper and making sure that the review process went well. Thanks to everyone involded for their commitment. 

On a more personal note, I have joined the list of JOSS reviewers and would be happy to help in future reviews. 

Best Regards, 

Miftari B
